### PR TITLE
4.0.0 Release

### PR DESCRIPTION
**Pull request by stanfordquan**
## Template 4.0.0 Release Notes

The three most important changes to the template as a part of this release are:
1. Compatibility with [gslab-python 4.0.0](https://github.com/gslab-econ/gslab_python).
2. New and improved pre-requisite testing.
3. We no longer version the `.sconsign.dblite`. This allows for users on different platforms to build fresh clones or new branches from the `cache` instead of having to rebuild the entire repository. See [here](https://github.com/gslab-econ/template/issues/40) and links therein for background. 

### Features
 - Uses [gslab-python 4.0.0](https://github.com/gslab-econ/gslab_python).
 - Moves [setup.py](https://github.com/gslab-econ/template/pull/52/files#diff-2eeaed663bd0d25b7e608891384b7298) to [configuration_test.py](https://github.com/gslab-econ/template/pull/52/files#diff-ab5ff5f3e09c044f03c3b74af0e50280) and improves pre-requisite testing.
 - Uses `atexit` to perform exit logging (see [here](https://github.com/gslab-econ/template/pull/52/files#diff-cc8093822b396591a5530a463c87d048R57)). 

### Bugs
 - We no longer version the [.sconsign.dblite](https://github.com/gslab-econ/template/pull/52/files#diff-4e28b2ade0be4c5b085ccbddfab81c23)

### Misc
 - Readme edits (see [here](https://github.com/gslab-econ/template/pull/52/files#diff-04c6e90faac2675aa89e2176d2eec7d8) and [here](https://github.com/gslab-econ/template/commit/6f1e5b046ebfdfb526a38c4f8320d45c039bcc13))
 - Switches to default metropolis beamer theme (see [here](https://github.com/gslab-econ/template/pull/52/files#diff-e9c57f7f78cdc25606eda2de68975854))
 - Small naming and formatting changes within [SConstruct](https://github.com/gslab-econ/template/pull/52/files#diff-cc8093822b396591a5530a463c87d048)

### CODE REVIEW

**Comment by lboxell** [comment]
@stanfordquan @arosenbe,

One thing to think about, per Toby, is whether we want to switch to using virtual environments (specifically via [conda](https://conda.io/docs/using/index.html)) for our template, either for 4.0 or to add to the outline.

**Comment by lboxell** [comment]
@arosenbe @stanfordquan, See my change log proposal below. We can't really have a "Here is what you need to do to update your template to be able to use gslab-python 4.0 list" because people/projects will be starting at different points, so we don't know what their "base template" version is. This issue is a reason why using virtual environments may be nice because we can tie a specific gslab-python version to each template version (and thus, people can continue to use older version of gslab-python easily with older versions of the template that require it).  


-----

## 4.0.0 Release Notes

The three most important changes to the template as a part of this release are:
1. Compatibility with [gslab-python 4.0.0](https://github.com/gslab-econ/gslab_python).
2. New and improved pre-requisite testing.
3. We no longer version the `.sconsign.dblite`. This allows for users on different platforms to build fresh clones or new branches from the `cache` instead of having to rebuild the entire repository. See [here](https://github.com/gslab-econ/template/issues/40) and links therein for background. 

### Features
 - Uses [gslab-python 4.0.0](https://github.com/gslab-econ/gslab_python).
 - Moves [setup.py](https://github.com/gslab-econ/template/pull/52/files#diff-2eeaed663bd0d25b7e608891384b7298) to [configuration_test.py](https://github.com/gslab-econ/template/pull/52/files#diff-ab5ff5f3e09c044f03c3b74af0e50280) and improves pre-requisite testing.
 - Uses `atexit` to perform exit logging (see [here](https://github.com/gslab-econ/template/pull/52/files#diff-cc8093822b396591a5530a463c87d048R57)). 

### Bugs
 - We no longer version the [.sconsign.dblite](https://github.com/gslab-econ/template/pull/52/files#diff-4e28b2ade0be4c5b085ccbddfab81c23)

### Misc
 - Readme edits (see [here](https://github.com/gslab-econ/template/pull/52/files#diff-04c6e90faac2675aa89e2176d2eec7d8))
 - Switches to default metropolis beamer theme (see [here](https://github.com/gslab-econ/template/pull/52/files#diff-e9c57f7f78cdc25606eda2de68975854))
 - Small naming and formatting changes within [SConstruct](https://github.com/gslab-econ/template/pull/52/files#diff-cc8093822b396591a5530a463c87d048)

**Comment by arosenbe** [comment]
@lboxell and @stanfordquan, 

Before this goes through, I think we should add two things to the FAQ in the readme: (i) how to pass command line arguments to scripts through builders and (ii) how to specify custom names for logs and the overwrites that will otherwise happen. I'm happy to write these, but wanted to get approval first.

**Comment by lboxell** [comment]
I have no qualms about that.

**Comment by arosenbe** [comment]
@stanfordquan we're missing https://github.com/gslab-econ/template/pull/52/commits/6f1e5b046ebfdfb526a38c4f8320d45c039bcc13 from the readme edits in the release notes.

**Comment by stanfordquan** [comment]
@gentzkow @jmshapir we prepared the release notes for template 4.0.0 [at the top of this pull request](https://github.com/gslab-econ/template/pull/52#issue-244890095). Please let us know what we can do to facilitate your review. Thanks!

Companion pull request at gslab_python [here](https://github.com/gslab-econ/gslab_python/pull/81).

**Comment by jmshapir** [comment]
@stanfordquan I am good with this if @gentzkow is!

**Comment by gentzkow** [comment]
I have reviewed these changes and parallel gslab_python [pull request](https://github.com/gslab-econ/gslab_python/issues/58#issuecomment-295959587). I think we should go ahead and merge / release this version.

I have a list of next steps I will make part of follow-up tasks following this release.

In addition, I have reviewed these changes alongside the lists / discussions of goals at the following locations:
* Comment thread to #19 
* RA notes in [gslab_python-#58](https://github.com/gslab-econ/gslab_python/issues/58#issuecomment-295959587)
* `admin` [outline](https://github.com/gslab-econ/admin/wiki/Outline)

The following are items we had decided to implement that have either not been done or are not documented clearly
1. Convention for pointing to local (e.g., raw data) directories external to the repository
2. Scons build step for checking out external files
3. "We should think about the templates we want R and Stata scripts to follow and implement those here. For Stata, this will mean revisiting our gslab-specific tools like preliminaries -- whether we want to keep using them and also how we want to install them."
4. Protocol for releases with large files that are not versioned
5. Where do issue markdown files fit in the template?
6. "I'd like to have a protocol that allows issue deliverables to be compiled pdfs including tables, figures, and other output, with clear provenance for where the output files came from. Simplest way would seem to be that we can release from an issue branch and in this case the released files are just the issue deliverables. Should think about how to make this work."
7. "gslab_stata should be installed to each local machine as a part of the computer build process and listed as a pre-req for any template that uses stata."

@stanfordquan: For each of these, why don't you let me know what the status is. For those we still want to implement we should push to the follow-up task. For those that need clearer documentation we can update the documentation on this release.

**Comment by gentzkow** [comment]
Addendum: 
8. We had previously discussed tracking Python and R dependencies so these can be installed automatically with e.g. the gslab_python library.

**Comment by stanfordquan** [comment]
@gentzkow 

1) `user_config.yaml` stores path to external resources, mentioned in repo's `README.md` alongside other prerequisites (since `user_config.yaml` is not versioned). Template's `README.md` can state that that is our suggested style with an example link. This convention applies to large files that have a local copy (e.g. `/Users/gentzkow/Dropbox/massivefile.txt`). 

2) For large files stored on cloud drives, [this issue](https://github.com/gslab-econ/gslab_python/issues/52#issuecomment-278149640) has the last words on the topic. 

'''
We are unsure whether we want to have the raw data download being part of the scons framework or a manual step that the user does, because we want scons to be error free as often as possible and the data download process had tended to be error prone in the past.
'''

In practice @arosenbe and I have been using `rclone` as part of our SCons build with few errors.

3) In `template`, we have removed R and Python scripts and greatly simplified the Stata script, so we currently do not have an official position on what those scripts should look like.

4) Everything meant to be released ends up in `/release`. Large files should be put into `.gitignore`. When you run `release.py`, the following happens:
	- The `/release` directory gets zipped and put on to cloud storage.
	- Everything versioned in the `/release` directory gets put on GitHub.

5) They don't have a spot in the `template`... see 6): 

6) The protocol right now is to produce issue markdown/pdf files in the branch, delete them before you merge into master, and copy paste a revision-stable link to the issue markdown in the concluding comment of an issue thread. The link should point to the final commit before merging, thus saving a snapshot of what the repo looks like when the issue deliverable was made. [The RA Manual currently reflects that.](https://github.com/gslab-econ/ra-manual/wiki/Tasks)

7) `gslab_stata` is not currently needed to run template. Do you want us to change that?

8) Tobias mentions using [`virtualenv`](https://virtualenv.pypa.io/en/stable/userguide/#usage). We haven't really looked into it.

**Comment by gentzkow** [comment]
Great. Thanks for the clarifications @stanfordquan.

Are the procedures for release you summarize in (4) documented somewhere? Let's make sure they are.

Otherwise, cleared to merge and close this issue.

**Comment by arosenbe** [comment]
I'm handling the updates to the readme discussed above.

File: SConstruct [hunk 1]
```diff
@@ -40,11 +46,13 @@ Default('./build', './release')
...
 if mode == 'cache':
-    CacheDir(cache_dir)
-
-# Print the state of the repo at end of SCons run
-finish_command = Command( 'state_of_repo.log', [], gslab_scons.misc.state_of_repo, MAXIT=10) # From http://stackoverflow.com/questions/8901296/how-do-i-run-some-code-after-every-build-in-scons
-Depends(finish_command, BUILD_TARGETS)
-env.AlwaysBuild(finish_command)
-if 'state_of_repo.log' not in BUILD_TARGETS: 
-    BUILD_TARGETS.append('state_of_repo.log')
+    (cache_dir)
```
Comment (lboxell) [inline SConstruct]: Does this need to be fixed?

File: SConstruct [hunk 2]
```diff
@@ -40,11 +46,13 @@ Default('./build', './release')
...
+
+debrief_env = {'MAXIT' : 10,
```
Comment (lboxell) [inline SConstruct]: Should we really be calling this 'env'? 'env' sort of have a special place in SCons.

File: SConstruct (hunk 1 above)
Comment (stanfordquan) [inline SConstruct]: That one really is a mystery to me. I fixed it back to `CacheDir(cache_dir)` a few times and for some reason it kept popping up. I'll fix it on `dev_four`.

File: SConstruct (hunk 2 above)
Comment (stanfordquan) [inline SConstruct]: Sure, how about `debrief_dict`?

Comment (stanfordquan) [inline SConstruct]: To be fair we pass this to `env = ` so it's not super unreasonable

Comment (lboxell) [inline SConstruct]: But is it really being used a Scons build step? I propose we change the arguments of https://github.com/gslab-econ/gslab_python/blob/dev_four/gslab_scons/misc.py#L14 also.

Comment (stanfordquan) [inline SConstruct]: Sure, let's do that.

Comment (lboxell) [inline SConstruct]: Ok, I made note in #56.

**Comment by stanfordquan** [comment]
(@arosenbe, @Shun-Yang, @yuchuan2016, @AakaashRao) I would appreciate it if we can do a quick test of `dev_four` template and gslab_python before the end of the day so that we can squash and merge both branches. Thanks!

**Comment by stanfordquan** [comment]
@arosenbe please also implement the `README.md` comments by MG above. Thanks.

**Comment by lboxell** [comment]
I tested and its fine up to my point about `CacheDir`.

**Comment by yuchuan2016** [comment]
I have run the template and have two comments:
- I need to manually delete my old version `gslab_make` etc in `site-packages`, or I will receive the `Your gslab_tools Python modules installation is outdated:` error.
- The first time I run `scons`, I will get a verbose `sconstruct.log` combining all `sconscript.log` together, like [this](https://github.com/gslab-econ/template/blob/3febe63fed90eabfcbf8246a7911e4cafc778ce1/sconstruct.log). Then I run `scons` for a second time, it only contains a few lines. Is this the intended feature that we combine all detailed `sconscript.log` of the current build at the root of the repo?

**Comment by stanfordquan** [comment]
@yuchuan2016 SConstruct.log appends SConscript.log files produced from that SCons run.

**Comment by arosenbe** [comment]
Readme updates discusses in https://github.com/gslab-econ/template/pull/52#issuecomment-318180875 and https://github.com/gslab-econ/template/pull/52#issuecomment-318294174 implemented at https://github.com/gslab-econ/template/commit/c1cdea3cbd4574d1506d42308eb3facc5481c52b.

**Comment by Shun-Yang** [comment]
@stanfordquan , I'm having the following error --- the lyx file in source/talk/ doesn't run.  Do you know what may be wrong? The lyx file in source/paper/ runs okay. 
```
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
build_lyx(["build/talk/slides.pdf"], ["source/talk/slides.lyx"])
Error: lyx did not run successfully.
       Please check that the executable, source, and target files
       Check SConstruct.log for errors.
Command tried: lyx -e pdf2 source/talk/slides.lyx > build/talk/sconscript.log
scons: *** [build/talk/slides.pdf] ExecCallError : 
Traceback (most recent call last):
  File "/usr/local/lib/scons-2.5.0/SCons/Action.py", line 1054, in execute
    result = self.execfunction(target=target, source=rsources, env=env)
  File "/Library/Python/2.7/site-packages/GSLab_Tools-4.0.0-py2.7.egg/gslab_scons/builders/build_lyx.py", line 58, in build_lyx
    raise ExecCallError(message)
ExecCallError
scons: building terminated because of errors.
_________________________________________________
```

**Comment by stanfordquan** [comment]
@Shun-Yang can you please post the full log? If the Lyx file didn't finish running, the `sconscript.log` will have been appended to the SConstruct.log.

**Comment by Shun-Yang** [comment]
@stanfordquan, the full log is here

<details>
<summary>log</summary>



*** New build: {2017-07-27 14:55:09} ***
scons: done reading SConscript files.
scons: Building targets ...
build_lyx(["build/talk/slides.pdf"], ["source/talk/slides.lyx"])
Error: lyx did not run successfully.
       Please check that the executable, source, and target files
       Check SConstruct.log for errors.
Command tried: lyx -e pdf2 source/talk/slides.lyx > build/talk/sconscript.log
scons: *** [build/talk/slides.pdf] ExecCallError : 
Traceback (most recent call last):
  File "/usr/local/lib/scons-2.5.0/SCons/Action.py", line 1054, in execute
    result = self.execfunction(target=target, source=rsources, env=env)
  File "/Library/Python/2.7/site-packages/GSLab_Tools-4.0.0-py2.7.egg/gslab_scons/builders/build_lyx.py", line 58, in build_lyx
    raise ExecCallError(message)
ExecCallError
scons: building terminated because of errors.
*** Build completed: {2017-07-27 14:55:10} ***
 
 
*** Warning!!! The log below does not have timestamps, the Sconscript may not have finished.
./build/talk/sconscript.log
This is pdfTeX, Version 3.14159265-2.6-1.40.16 (TeX Live 2015) (preloaded format=pdflatex)
 restricted \write18 enabled.
entering extended mode
(./slides.tex
LaTeX2e <2015/01/01>
Babel <3.9l> and hyphenation patterns for 79 languages loaded.

</details>

**Comment by stanfordquan** [comment]
Yeah Lyx often nopes out without an informative message. I would try to compile the `slides.lyx` file from within Lyx and see if that runs.

**Comment by Shun-Yang** [comment]
@stanfordquan thanks! I think it's because my beamer_gslab theme didn't install correctly.  The repo runs fine on my end.

**Comment by arosenbe** [comment]
Standard use cases all error free for me 🎉

**Comment by stanfordquan** [comment]
@Shun-Yang we don't need `beamer_gslab` to run `template.`

**Comment by stanfordquan** [comment]
Is your template on `dev_four`?

**Comment by Shun-Yang** [comment]
@stanfordquan I'm on `dev_four`. Sorry for the confusion.  I think it's because `beamer_metropolis` is not installed in my computer. I have `beamer_metropolis_gslab` installed.

**Comment by stanfordquan** [comment]
Cool, @Shun-Yang ."Metropolis is on CTAN since December 2014 thus it is part of MikTeX and will be part of TeX Live 2016." Is there any issue other than this?

**Comment by Shun-Yang** [comment]
@stanfordquan not sure why the theme isn't on my laptop. I installed it and the repo runs good on my end. No other issue!

**Comment by stanfordquan** [comment]
Perfect, that's a wrap!

**Comment by stanfordquan** [comment]
See release [here](https://github.com/gslab-econ/template/tree/v4.0.0).

**Comment by Shun-Yang** [comment]
Quick question here: for existing repo (e.g. ad-price-drivers), is the protocol to switch to new template with gslab_python 4.0.0 or stick to the original one and don't update gslab_python

**Comment by jmshapir** [comment]
If @gentzkow has a preference let's go with that; I would suggest we do whatever is more convenient.

**Comment by gentzkow** [comment]
I think the expectation should be that we will only have the most recent version of gslab_python installed. So any repos that we are actively using should be upgraded to conform to the current version of gslab_python. We don't necessarily have to import all changes from `template` at the same time, but that may be the easiest thing to do in most cases.

**Comment by jmshapir** [comment]
@stanfordquan @gentzkow I don't have a good sense of how closely tied gslab_python is to the directory structure. My sense is that they're not that tightly related.

If they are, then I would hesitate somewhat. I think it would be pretty costly to have to reorganize all the file paths and structure of each repo every time we release gslab_python. But again I'm assuming it shouldn't be quite that extensive.

**Comment by gentzkow** [comment]
Right. I'm assuming nothing about the directory structure is hard-coded in the Python library. (I wouldn't think there'd be any reason for that.) I also think that we should be settling down to a point where subsequent releases can all be backwards-compatible so we don't need to do this very often.

**Pull request merged**